### PR TITLE
 Calculate expires_at from issued_at, if provided

### PR DESCRIPTION
**Pull request by lestopher**
This pull request fixes a subtle issue where if a user provides `expires_in` and `issued_at`, the `expires_at` is calculated with `Time.now`, causing the client to never be expired.

Examples:
Current result
```ruby
# Assuming the current time is 2018-08-24T12:00:00
client.update_token!(
  expires_in: 3600,
  issued_at: '2018-08-24T00:00:00Z'
)
client.expires_at # => 2018-08-24T13:00:00
client.expired? # => false
```

Expected result
```ruby
# Assuming the current time is 2018-08-24T12:00:00
client.update_token!(
  expires_in: 3600,
  issued_at: '2018-08-24T00:00:00Z'
)
client.expires_at # => 2018-08-24T01:00:00
client.expired? # => true
```

Notice that `self.issued_at =` is moved above the expires code block, as `expires_in=` checks if `self.issued_at` is set or not.

### CODE REVIEW

**Comment by lestopher** [comment]
I signed it!

**Comment by dazuma** [comment]
The problem looks real, but I think the fix needs to be elsewhere. The current proposal would cause:

```ruby
# Assuming the current time is 2018-08-24T12:00:00
client.issued_at # => 2018-08-24T00:00:00
client.expires_in = 3600
client.expires_at # => 2018-08-24T01:00:00
# But we expect it to "expire in an hour", or at 2018-08-24T13:00:00
```

I think we need to fix this in `update_token!` by having special logic for setting both `expires_in` and `issued_at` simultaneously.

**Comment by lestopher** [comment]
I was under the impression that the `expires_in` field was relative to `issued_at` field if it was provided. Is it always relative to the current time?

I've always treated `issued_at` as the time when the access token was returned in the access token response.

**Comment by TheRoyalTnetennba** [comment]
`expires_in` is relative to `issued_at`. I agree with the general thrust of this fix, but I am curious, when would `expires_in` be updated at a time other than `issued_at`?

**Comment by lestopher** [comment]
That's a good question, I don't know what other reference time it'd be compared to other than `issued_at`.

**Comment by foxtacles** [comment]
This PR would resolve #116, fixing the breaking change introduced with version 0.9.0. Merge would be much appreciated 👍 

I personally wouldn't expect setting `expires_in` to cause `issued_at` being reset to `Time.now`, so this fix seems perfectly appropriate to me.

**Comment by foxtacles** [comment]
Ah, I didn't realize that. Then I agree it should be handled in `update_token!`.

**Comment by lestopher** [comment]
Updated to keep the original `expires_in` behavior, and a special case within `updated_token!` to handle both fields being set at once.

File: lib/signet/oauth_2/client.rb [hunk 1]
```diff
@@ -1196,6 +1201,14 @@ def normalize_timestamp(time)
           fail "Invalid time value #{time}"
         end
       end
+
+      def set_relative_expires_at(issued_at, expires_in)
+        self.expires_in = expires_in.to_i
```
Comment (dazuma) [inline lib/signet/oauth_2/client.rb]: You can remove this line. Both `expires_in=` and `expires_at=` set the same state.

File: lib/signet/oauth_2/client.rb [hunk 2]
```diff
@@ -743,7 +748,7 @@ def expires_in
       #   The access token lifetime.
       def expires_in=(new_expires_in)
         if new_expires_in != nil
-          @issued_at = Time.now
+          @issued_at = normalize_timestamp(Time.now)
```
Comment (dazuma) [inline lib/signet/oauth_2/client.rb]: One more thing: normalize_timestamp should not be necessary here, because it's a noop for instances of Time.

Comment (lestopher) [reply lib/signet/oauth_2/client.rb]: Ah gotcha, should have looked at the implementation.

**Comment by dazuma** [comment]
Thanks @lestopher ! Sorry this took so long to get reviewed.

**Comment by dazuma** [comment]
Will get a release out shortly.

**Comment by lestopher** [comment]
No problem, thanks for the background and context. I wasn't entirely sure why we needed to keep the standard behavior (from the first comment), but the conversation helped. Glad we can get this fixed for everybody else.

**Pull request merged**